File: src/client/cursor.ts
/**
 * FindCursor - MongoDB-compatible cursor for query results
 *
 * Provides chainable methods for sorting, limiting, skipping, and projecting
 * query results. Supports async iteration and array conversion.
 */

import type { MongoCollection } from './mongo-collection'

// Base document type
export type Document = Record<string, unknown>

export interface FindOptions {
  projection?: Record<string, 0 | 1>
  sort?: Record<string, 1 | -1>
  limit?: number
  skip?: number
}

/**
 * FindCursor provides lazy evaluation of MongoDB queries
 */
export class FindCursor<TSchema extends Document = Document> {
  private readonly collection: MongoCollection<any>
  private readonly filter: object
  private _projection?: Record<string, 0 | 1>
  private _sort?: Record<string, 1 | -1>
  private _limit?: number
  private _skip?: number
  private _executed: boolean = false
  private _results: TSchema[] = []

  constructor(
    collection: MongoCollection<any>,
    filter: object,
    options?: FindOptions
  ) {
    this.collection = collection
    this.filter = filter
    if (options) {
      if (options.projection !== undefined) this._projection = options.projection
      if (options.sort !== undefined) this._sort = options.sort
      if (options.limit !== undefined) this._limit = options.limit
      if (options.skip !== undefined) this._skip = options.skip
    }
  }

  /**
   * Set the sort order
   */
  sort(spec: Record<string, 1 | -1>): this {
    this._sort = spec
    return this
  }

  /**
   * Set the maximum number of documents to return
   */
  limit(count: number): this {
    this._limit = count
    return this
  }

  /**
   * Set the number of documents to skip
   */
  skip(count: number): this {
    this._skip = count
    return this
  }

  /**
   * Set the projection for returned documents
   */
  project(spec: Record<string, 0 | 1>): this {
    this._projection = spec
    return this
  }

  /**
   * Execute the query and return all documents as an array
   */
  async toArray(): Promise<TSchema[]> {
    if (!this._executed) {
      const options: { sort?: Record<string, 1 | -1>; skip?: number; limit?: number } = {}
      if (this._sort !== undefined) options.sort = this._sort
      if (this._skip !== undefined) options.skip = this._skip
      if (this._limit !== undefined) options.limit = this._limit
      this._results = this.collection._findDocuments(this.filter, options)

      // Apply projection
      if (this._projection) {
        this._results = this._results.map(doc =>
          this.collection._applyProjection(doc, this._projection!)
        )
      }

      this._executed = true
    }
    return [...this._results]
  }

  /**
   * Execute callback for each document
   */
  async forEach(callback: (doc: TSchema) => void): Promise<void> {
    const docs = await this.toArray()
    for (const doc of docs) {
      callback(doc)
    }
  }

  /**
   * Count documents matching the query
   */
  async count(): Promise<number> {
    const docs = await this.toArray()
    return docs.length
  }

  /**
   * Check if there are any documents
   */
  async hasNext(): Promise<boolean> {
    const docs = await this.toArray()
    return docs.length > 0
  }

  /**
   * Get the next document (simple implementation)
   */
  async next(): Promise<TSchema | null> {
    const docs = await this.toArray()
    return docs.length > 0 ? docs[0] ?? null : null
  }

  /**
   * Map over documents
   */
  map<U>(fn: (doc: TSchema) => U): MappedCursor<TSchema, U> {
    return new MappedCursor(this, fn)
  }

  /**
   * Async iterator support
   */
  async *[Symbol.asyncIterator](): AsyncIterableIterator<TSchema> {
    const docs = await this.toArray()
    for (const doc of docs) {
      yield doc
    }
  }
}

/**
 * Cursor with map transformation applied
 */
class MappedCursor<TSchema extends Document, U> {
  private readonly cursor: FindCursor<TSchema>
  private readonly mapFn: (doc: TSchema) => U

  constructor(cursor: FindCursor<TSchema>, mapFn: (doc: TSchema) => U) {
    this.cursor = cursor
    this.mapFn = mapFn
  }

  async toArray(): Promise<U[]> {
    const docs = await this.cursor.toArray()
    return docs.map(this.mapFn)
  }

  async forEach(callback: (doc: U) => void): Promise<void> {
    const docs = await this.toArray()
    for (const doc of docs) {
      callback(doc)
    }
  }

  async *[Symbol.asyncIterator](): AsyncIterableIterator<U> {
    const docs = await this.toArray()
    for (const doc of docs) {
      yield doc
    }
  }
}

export default FindCursor

File: sdks/python/src/mongo_do/cursor.py
"""
Cursor - Async cursor for iterating over query results.

Provides a PyMongo-compatible cursor interface for streaming
documents from MongoDB queries over RPC.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, AsyncIterator, Generic, TypeVar

if TYPE_CHECKING:
    from rpc_do import RpcClient

    from .types import Filter, Projection, Sort

T = TypeVar("T", bound=dict[str, Any])

__all__ = ["Cursor"]


class Cursor(Generic[T]):
    """
    Async cursor for iterating over query results.

    Cursor provides a lazy, async iterable interface for MongoDB query
    results. It supports chaining operations like sort, limit, skip,
    and projection before iteration begins.

    Example:
        async for doc in collection.find({"status": "active"}):
            print(doc)

        # With chaining
        cursor = collection.find({}).sort("created_at", -1).limit(10)
        async for doc in cursor:
            print(doc)
    """

    __slots__ = (
        "_rpc",
        "_database",
        "_collection",
        "_filter",
        "_projection",
        "_sort",
        "_limit",
        "_skip",
        "_batch_size",
        "_results",
        "_exhausted",
        "_position",
    )

    def __init__(
        self,
        rpc: RpcClient,
        database: str,
        collection: str,
        filter: Filter | None = None,
        projection: Projection = None,
    ) -> None:
        """
        Initialize a cursor.

        Args:
            rpc: The RPC client for making calls.
            database: Database name.
            collection: Collection name.
            filter: Query filter.
            projection: Fields to include/exclude.
        """
        self._rpc = rpc
        self._database = database
        self._collection = collection
        self._filter: Filter = filter or {}
        self._projection: Projection = projection
        self._sort: Sort = None
        self._limit: int = 0
        self._skip: int = 0
        self._batch_size: int = 100
        self._results: list[T] | None = None
        self._exhausted: bool = False
        self._position: int = 0

    def sort(self, key_or_list: str | list[tuple[str, int]], direction: int = 1) -> Cursor[T]:
        """
        Sort the results.

        Args:
            key_or_list: Field name or list of (field, direction) tuples.
            direction: Sort direction (1 for ascending, -1 for descending).
                       Only used if key_or_list is a string.

        Returns:
            Self for chaining.
        """
        if isinstance(key_or_list, str):
            self._sort = [(key_or_list, direction)]
        else:
            self._sort = key_or_list
        return self

    def limit(self, limit: int) -> Cursor[T]:
        """
        Limit the number of results.

        Args:
            limit: Maximum number of documents to return.

        Returns:
            Self for chaining.
        """
        self._limit = limit
        return self

    def skip(self, skip: int) -> Cursor[T]:
        """
        Skip the first N results.

        Args:
            skip: Number of documents to skip.

        Returns:
            Self for chaining.
        """
        self._skip = skip
        return self

    def batch_size(self, size: int) -> Cursor[T]:
        """
        Set the batch size for fetching results.

        Args:
            size: Number of documents per batch.

        Returns:
            Self for chaining.
        """
        self._batch_size = size
        return self

    def project(self, projection: Projection) -> Cursor[T]:
        """
        Set field projection.

        Args:
            projection: Fields to include/exclude.

        Returns:
            Self for chaining.
        """
        self._projection = projection
        return self

    async def _execute(self) -> list[T]:
        """
        Execute the query and fetch results.

        Returns:
            List of documents matching the query.
        """
        if self._results is not None:
            return self._results

        # Build the query options
        options: dict[str, Any] = {}

        if self._projection:
            if isinstance(self._projection, list):
                # Convert list of field names to projection dict
                options["projection"] = {field: 1 for field in self._projection}
            else:
                options["projection"] = dict(self._projection)

        if self._sort:
            options["sort"] = self._sort

        if self._limit > 0:
            options["limit"] = self._limit

        if self._skip > 0:
            options["skip"] = self._skip

        # Execute via RPC
        result = await self._rpc.mongo.find(
            self._database,
            self._collection,
            self._filter,
            options,
        )

        self._results = result if isinstance(result, list) else []
        return self._results

    async def to_list(self, length: int | None = None) -> list[T]:
        """
        Convert cursor to a list.

        Args:
            length: Maximum number of documents to return.
                    If None, returns all documents.

        Returns:
            List of documents.
        """
        results = await self._execute()
        if length is not None:
            return results[:length]
        return results

    def __aiter__(self) -> AsyncIterator[T]:
        """Return async iterator."""
        return self

    async def __anext__(self) -> T:
        """
        Get the next document.

        Returns:
            The next document.

        Raises:
            StopAsyncIteration: When all documents have been iterated.
        """
        # Lazy execute on first iteration
        if self._results is None:
            await self._execute()

        assert self._results is not None

        if self._position >= len(self._results):
            self._exhausted = True
            raise StopAsyncIteration

        doc = self._results[self._position]
        self._position += 1
        return doc

    async def next(self) -> T:
        """
        Get the next document.

        Returns:
            The next document.

        Raises:
            StopAsyncIteration: When all documents have been iterated.
        """
        return await self.__anext__()

    async def count(self) -> int:
        """
        Count documents matching the query.

        Returns:
            Number of documents.

        Note:
            This method is deprecated in PyMongo 4.0+.
            Use count_documents() on the collection instead.
        """
        result = await self._rpc.mongo.countDocuments(
            self._database,
            self._collection,
            self._filter,
        )
        return result if isinstance(result, int) else 0

    async def distinct(self, key: str) -> list[Any]:
        """
        Get distinct values for a field.

        Args:
            key: Field name to get distinct values for.

        Returns:
            List of distinct values.
        """
        result = await self._rpc.mongo.distinct(
            self._database,
            self._collection,
            key,
            self._filter,
        )
        return result if isinstance(result, list) else []

    def clone(self) -> Cursor[T]:
        """
        Clone this cursor.

        Returns:
            A new cursor with the same query parameters.
        """
        cursor = Cursor[T](
            self._rpc,
            self._database,
            self._collection,
            self._filter,
            self._projection,
        )
        cursor._sort = self._sort
        cursor._limit = self._limit
        cursor._skip = self._skip
        cursor._batch_size = self._batch_size
        return cursor

    @property
    def alive(self) -> bool:
        """Check if the cursor can still yield documents."""
        return not self._exhausted

    def rewind(self) -> Cursor[T]:
        """
        Rewind the cursor to the beginning.

        Returns:
            Self for chaining.
        """
        self._position = 0
        self._exhausted = False
        return self
